cmd/solana_exporter/slots.go

```diff
@@ -68,182 +87,161 @@ func init() {
+		assertf(
+			firstSlot == c.lastSlot+1,
+			"first slot %v does not follow from current last slot %v",
+			firstSlot,
+			c.lastSlot,
+		)
 
-	return info.Epoch, firstSlot, lastSlot
-}
+		// and also, make sure that we have completed the last epoch:
+		assertf(
+			c.slotWatermark == c.lastSlot,
+			"can't update epoch when watermark %v hasn't reached current last-slot %v",
+			c.slotWatermark,
+			c.lastSlot,
+		)
 
-func (c *solanaCollector) updateCounters(epoch, firstSlot int64, lastSlotOpt *int64) (int64, error) {
-	ctx, cancel := context.WithTimeout(context.Background(), httpTimeout)
-	defer cancel()
+		// the epoch number is progressing correctly, so we can update our tracking numbers:
+		c.currentEpoch = epoch.Epoch
+		c.firstSlot = firstSlot
+		c.lastSlot = lastSlot
+	}
 
-	var lastSlot int64
-	var err error
+	// emit epoch bounds:
+	currentEpochNumber.Set(float64(c.currentEpoch))
+	epochFirstSlot.Set(float64(c.firstSlot))
+	epochLastSlot.Set(float64(c.lastSlot))
+}
 
-	if lastSlotOpt == nil {
-		lastSlot, err = c.rpcClient.GetSlot(ctx)
+// closeCurrentEpoch is called when an epoch change-over happens, and we need to make sure we track the last
+// remaining slots in the "current" epoch before we start tracking the new one.
+func (c *SlotWatcher) closeCurrentEpoch(ctx context.Context, newEpoch *rpc.EpochInfo) {
+	c.fetchAndEmitBlockProduction(ctx, c.lastSlot)
+	c.trackEpoch(newEpoch)
+}
 
-		if err != nil {
-			return 0, fmt.Errorf("error while getting the last slot: %v", err)
-		}
-		klog.V(2).Infof("Setting lastSlot to %d", lastSlot)
-	} else {
-		lastSlot = *lastSlotOpt
-		klog.Infof("Got lastSlot: %d", lastSlot)
+// checkValidSlotRange makes sure that the slot range we are going to query is within the current epoch we are tracking.
+func (c *SlotWatcher) checkValidSlotRange(from, to int64) error {
+	if from < c.firstSlot || to > c.lastSlot {
+		return fmt.Errorf(
+			"start-end slots (%v -> %v) is not contained within current epoch %v range (%v -> %v)",
+			from,
+			to,
+			c.currentEpoch,
+			c.firstSlot,
+			c.lastSlot,
+		)
 	}
+	return nil
+}
 
-	if firstSlot > lastSlot {
-		return 0, fmt.Errorf(
-			"in epoch %d, firstSlot (%d) > lastSlot (%d), this should not happen, not updating",
-			epoch,
-			firstSlot,
-			lastSlot,
-		)
+// fetchAndEmitBlockProduction fetches block production up to the provided endSlot, emits the prometheus metrics,
+// and updates the SlotWatcher.slotWatermark accordingly
+func (c *SlotWatcher) fetchAndEmitBlockProduction(ctx context.Context, endSlot int64) {
+	// add 1 because GetBlockProduction's range is inclusive, and the watermark is already tracked
+	startSlot := c.slotWatermark + 1
+	klog.Infof("Fetching block production in [%v -> %v]", startSlot, endSlot)
+
+	// make sure the bounds are contained within the epoch we are currently watching:
+	if err := c.checkValidSlotRange(startSlot, endSlot); err != nil {
+		panic(err)
```

Comment:
Is a panic really necessary in this case? Can we not just gracefully skip this or reset things some how? Also, perhaps instead of a panic (which dumps all goroutines + stack traces to stdout, we just do something more like [klog.Fatal(err.String())](https://pkg.go.dev/k8s.io/klog/v2#Fatal) or maybe [klog.Fatalf("slot range is out of bounds of the current epoch: %v", err)](https://pkg.go.dev/k8s.io/klog/v2#Fatalf)

Reply:
The reason I panic here is because if it happens then it's a logic issue in the code, not just some random event.